### ISSUE TITLE
Does "retrain" from last checkpoint take into account the "best.pt" weights?

### ISSUE THREAD

**Issue by hdnh2006** [ultralytics/yolov5#2182]
## ❔Question
My computer has exploded for some Ubuntu 20.04 bug while I was training a model with a large dataset:
![image](https://user-images.githubusercontent.com/17271049/107620284-bbdfef80-6c54-11eb-87ab-887cbd9a37bd.png)

Fortunately, with this wonderful code that Glenn has created you can continue training from last checkpoint using the argument `weights` in `train.py` like following: `python train.py --weights yourproject/last.pt ...`

That's good, but **my question is:** does this "retrain" take in account the "best.pt" weights? I mean, how the code knows which are the "best.pt" weights? I think for "retrain" I have to pass as argument the "best.pt" weights, but I am not pretty sure.

Can somebody give me information about this?

**Comment by NanoCode012** [comment]
@hdnh2006 , you can specify the best weights with `--resume path/to/weights.pt`. If you don't specify best weights, it will take the latest `last.pt`.
https://github.com/ultralytics/yolov5/blob/404749a33cc29d119f54b2ce35bf3b33a847a487/train.py#L449
https://github.com/ultralytics/yolov5/blob/404749a33cc29d119f54b2ce35bf3b33a847a487/train.py#L481-L489
https://github.com/ultralytics/yolov5/blob/404749a33cc29d119f54b2ce35bf3b33a847a487/utils/general.py#L44-L47

**Comment by glenn-jocher** [comment]
@hdnh2006 at the moment we differentiate between --resume for resuming an **interrupted** run using all the same settings/hyps, and simply starting a training from pretrained --weights. You should only resume from last.pt if resuming. When **starting** a new training from pretrained weights, you can select any weights you want.

There are two options to resume an interrupted run are:
```bash
python train.py --resume  # resume latest training
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

If you started the training with a multi-GPU command then you must resume it with the same exact configuration (and vice versa). Multi-GPU resume commands are here, assuming you are using 8 GPUs:
```bash
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume  # resume latest training
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

Note that you may not change settings when resuming, you must resume with the same exact settings (`--epochs`, `--batch` etc).

**Comment by hdnh2006** [comment]
Thanks @NanoCode012, @glenn-jocher for the quick answer. 

Let me see if I understood, if y use the `--resume` parameter, will it consider the "best.pt" weights that are in the same folder than "last.pt" weights?

**Comment by glenn-jocher** [comment]
@hdnh2006 `--resume` automatically finds your most recently saved last.pt file.

**Comment by hdnh2006** [comment]
Thanks @glenn-jocher, my problem is not with last.pt file but with best.pt files which are at the end the most important weights, are those weights considered if I use the `--resume` parameter? 

Thanks Glenn.

**Comment by glenn-jocher** [comment]
@hdnh2006 no, --resume will automatically find your most recent last.pt file. You can instead point to any --resume checkpoint you want, including best.pt files, provided the run was interrupted and training not completed.

**Comment by hdnh2006** [comment]
That's what I did it but in the new "exp" folder I didn't see a new "best.pt" file, so I think in this case it just saves the last.pt weight, am I right?

**Comment by glenn-jocher** [comment]
best.pt files will always be present unless you have used the --nosave argument during your training.

**Comment by shreyash9806** [comment]
I have used this command to resume my training but no best.pt files are saved in runs/exp
python train.py --data trash.yaml --cfg yolov5x.yaml --weights '' --batch-size 6

**Comment by shreyash9806** [comment]
can

> best.pt files will always be present unless you have used the --nosave argument during your training.

can you help me what i am doing wrong

**Comment by glenn-jocher** [comment]
@shreyash9806 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/aa542ce6a65658ff931fee9bbab77c0145c152f0/train.py#L175-L181) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="Screenshot 2022-04-10 at 11 11 51" src="https://user-images.githubusercontent.com/26833433/162611202-4e0b611f-6cca-4e43-94fc-bbc44e6447f8.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by gshoanganh** [comment]
hi!
Do you have any way to set new EPOCHS, BATCH?

**Comment by glenn-jocher** [comment]
@gshoanganh hello! Yes, you can modify the number of epochs and batch size in YOLOv5. To change the number of epochs, you can set the `--epochs` argument when starting the training script. For example: `python train.py --epochs 200` will train the model for 200 epochs.

To modify the batch size, you can set the `--batch-size` argument when starting the training script. For example: `python train.py --batch-size 16` will use a batch size of 16.

Remember to choose appropriate values for these parameters based on your hardware resources and dataset characteristics. Let me know if you have any other questions!
